hibernate-core/src/main/java/org/hibernate/sql/results/internal/domain/CircularBiDirectionalFetchImpl.java

```diff
@@ -182,11 +181,13 @@ public Object assemble(RowProcessingState rowProcessingState, JdbcValuesSourcePr
 			final Object initializedInstance = initializer.getInitializedInstance();
 			final LazyInitializer lazyInitializer = HibernateProxy.extractLazyInitializer( initializedInstance );
 			if ( lazyInitializer != null ) {
-				if ( initializedInstance.getClass().isAssignableFrom( javaType.getJavaTypeClass() ) ) {
-					return initializedInstance;
-				}
```

Comment:
Shouldn't we still do this if the circular association is marked as LAZY?

Reply:
not sure I get your comment, by the way this issue seems happening only when we have an EAGER association

Reply:
With your change, you are now always initializing and narrowing the proxy, but shouldn't we only do this when the association should be fetched eagerly?

I think you should re-add the following:

```java
				if ( timing == LAZY && initializedInstance.getClass().isAssignableFrom( javaType.getJavaTypeClass() ) ) {
					return initializedInstance;
				}
```